File: _archive_ts/packages/core/src/policy/router.ts
/**
 * 策略路由器（OpenHarness 二开版）
 * 作用：把“输入上下文”映射为“可执行路由决策”，实现可审计的规则命中链路。
 */

import type {
  PolicyCondition,
  PolicyLeafCondition,
  RoutingContext,
  RoutingDecision,
  RoutingPolicy,
  RoutingRule,
} from '../types.js';

/**
 * 路由器：根据策略配置执行规则匹配，输出标准化 RoutingDecision。
 */
export class PolicyRouter {
  constructor(private readonly policy: RoutingPolicy) {}

  /**
   * 执行一次路由决策。
   * 为什么这样做：统一把 rule match、fallback、trace 封装为单入口，便于审计与测试。
   */
  route(input: RoutingContext): RoutingDecision {
    const sortedRules = [...this.policy.rules].sort((a, b) => b.priority - a.priority);
    const evaluatedRules: string[] = [];

    let selectedRule: RoutingRule | null = null;
    for (const rule of sortedRules) {
      evaluatedRules.push(rule.id);
      if (this.matches(rule.when, input)) {
        selectedRule = rule;
        break;
      }
    }

    const selectedRoute = selectedRule?.route ?? this.policy.fallback;
    const strategy = this.policy.strategies[selectedRoute.strategy];
    if (!strategy) {
      throw new Error(`Routing strategy not found: ${selectedRoute.strategy}`);
    }

    const stopProfile = this.policy.stopProfiles[selectedRoute.stopProfile];
    if (!stopProfile) {
      throw new Error(`Stop profile not found: ${selectedRoute.stopProfile}`);
    }

    const confidence = this.computeConfidence(selectedRule, sortedRules.length);
    const selectedRuleId = selectedRule?.id ?? 'FALLBACK';

    return {
      selectedStrategy: selectedRoute.strategy,
      selectedRoute: {
        mode: strategy.mode,
        runtime: strategy.runtime,
        autonomy: strategy.autonomy,
        score: confidence,
      },
      stopProfile: selectedRoute.stopProfile,
      activeStopConditions: stopProfile.conditionIds,
      reasonCodes: [selectedRuleId, selectedRoute.reason],
      requiredCapabilities: strategy.requiredCapabilities,
      trace: {
        evaluatedRules,
        selectedRule: selectedRuleId,
        timestamp: input.timestamp,
      },
    };
  }

  /**
   * 判断条件是否命中。
   * 为什么这样做：支持 all/any 组合条件，便于把路由规则从代码迁移到配置。
   */
  private matches(condition: PolicyCondition, input: RoutingContext): boolean {
    if ('all' in condition || 'any' in condition) {
      const group = condition;
      if (group.all && group.all.length > 0) {
        return group.all.every((item) => this.matches(item, input));
      }
      if (group.any && group.any.length > 0) {
        return group.any.some((item) => this.matches(item, input));
      }
      return false;
    }

    const leaf = condition as PolicyLeafCondition;
    const actual = this.readPath(input, leaf.field);
    return this.compare(actual, leaf.op, leaf.value);
  }

  /**
   * 读取点路径字段。
   * 为什么这样做：规则字段使用字符串路径，路由器必须通用读取不同输入结构。
   */
  private readPath(data: unknown, path: string): unknown {
    const segments = path.split('.');
    let cursor: unknown = data;
    for (const segment of segments) {
      if (!cursor || typeof cursor !== 'object' || !(segment in cursor)) {
        return undefined;
      }
      cursor = (cursor as Record<string, unknown>)[segment];
    }
    return cursor;
  }

  /**
   * 执行操作符比较。
   * 为什么这样做：把规则表达式转换为统一比较逻辑，便于扩展新操作符。
   */
  private compare(actual: unknown, op: PolicyLeafCondition['op'], expected: unknown): boolean {
    switch (op) {
      case 'eq':
        return actual === expected;
      case 'ne':
        return actual !== expected;
      case 'gt':
        return typeof actual === 'number' && typeof expected === 'number' && actual > expected;
      case 'gte':
        return typeof actual === 'number' && typeof expected === 'number' && actual >= expected;
      case 'lt':
        return typeof actual === 'number' && typeof expected === 'number' && actual < expected;
      case 'lte':
        return typeof actual === 'number' && typeof expected === 'number' && actual <= expected;
      case 'in':
        return Array.isArray(expected) && expected.includes(actual);
      case 'not_in':
        return Array.isArray(expected) && !expected.includes(actual);
      default:
        return false;
    }
  }

  /**
   * 计算路由置信度。
   * 为什么这样做：命中越靠前优先级越高，给更高分用于下游审计分析。
   */
  private computeConfidence(rule: RoutingRule | null, totalRules: number): number {
    if (!rule || totalRules <= 0) return 0.55;
    const sorted = [...this.policy.rules].sort((a, b) => b.priority - a.priority);
    const index = sorted.findIndex((item) => item.id === rule.id);
    if (index < 0) return 0.6;
    const normalized = 1 - index / Math.max(totalRules, 1);
    return Math.round((0.6 + normalized * 0.4) * 1000) / 1000;
  }
}

File: src/velaris_agent/velaris/router.py
"""Velaris 原生策略路由器。"""

from __future__ import annotations

from dataclasses import asdict, dataclass
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

import yaml

from velaris_agent.scenarios.registry import ScenarioRegistry


@dataclass(frozen=True)
class SelectedRoute:
    """标准化路由结果。"""

    mode: str
    runtime: str
    autonomy: str
    score: float


@dataclass(frozen=True)
class RoutingDecision:
    """路由决策结果。"""

    selected_strategy: str
    selected_route: SelectedRoute
    stop_profile: str
    active_stop_conditions: list[str]
    reason_codes: list[str]
    required_capabilities: list[str]
    trace: dict[str, Any]
    governance: dict[str, Any]

    def to_dict(self) -> dict[str, Any]:
        """把路由决策转换为 JSON 友好的字典。"""
        return asdict(self)


class PolicyRouter:
    """Velaris 原生策略路由器。"""

    def __init__(self, policy_path: str | Path | None = None, scenario_registry: ScenarioRegistry | None = None) -> None:
        self.policy_path = Path(policy_path).resolve() if policy_path is not None else self._resolve_default_policy_path()
        self.policy = self._load_policy(self.policy_path)
        self._registry = scenario_registry or ScenarioRegistry()

    def route(self, plan: dict[str, Any], query: str) -> RoutingDecision:
        routing_context = self._build_routing_context(plan=plan, query=query)
        sorted_rules = sorted(self.policy["rules"], key=lambda item: int(item["priority"]), reverse=True)
        evaluated_rules: list[str] = []
        selected_rule: dict[str, Any] | None = None

        for rule in sorted_rules:
            evaluated_rules.append(str(rule["id"]))
            if self._matches(rule["when"], routing_context):
                selected_rule = rule
                break

        route_config = selected_rule["route"] if selected_rule is not None else self.policy["fallback"]
        selected_rule_id = str(selected_rule["id"]) if selected_rule is not None else "FALLBACK"
        strategy_name = str(route_config["strategy"])
        strategy = self.policy["strategies"][strategy_name]
        stop_profile_name = str(route_config["stop_profile"])
        stop_profile = self.policy["stop_profiles"][stop_profile_name]

        required_capabilities = list(dict.fromkeys(strategy.get("required_capabilities", [])))
        confidence = self._compute_confidence(selected_rule=selected_rule, sorted_rules=sorted_rules)

        return RoutingDecision(
            selected_strategy=strategy_name,
            selected_route=SelectedRoute(
                mode=str(strategy["mode"]),
                runtime=str(strategy["runtime"]),
                autonomy=str(strategy["autonomy"]),
                score=confidence,
            ),
            stop_profile=stop_profile_name,
            active_stop_conditions=[str(condition["id"]) for condition in stop_profile.get("conditions", [])],
            reason_codes=[
                selected_rule_id,
                str(route_config.get("reason", self.policy["fallback"].get("reason", "fallback"))),
                str(plan.get("scenario", "general")),
            ],
            required_capabilities=required_capabilities,
            trace={
                "evaluated_rules": evaluated_rules,
                "selected_rule": selected_rule_id,
                "timestamp": datetime.now(UTC).isoformat(),
                "policy_id": self.policy.get("policy_id"),
                "routing_context": routing_context,
            },
            governance=dict(plan.get("governance", {})),
        )

    def _resolve_default_policy_path(self) -> Path:
        """多策略查找 config/routing-policy.yaml: CWD 优先, __file__ 回退。"""
        # 策略1: 从当前工作目录向上查找
        cwd = Path.cwd().resolve()
        for parent in [cwd, *cwd.parents]:
            candidate = parent / "config" / "routing-policy.yaml"
            if candidate.exists():
                return candidate
        # 策略2: 从 __file__ 所在目录向上查找 (原有逻辑, 作为回退)
        current = Path(__file__).resolve()
        for parent in current.parents:
            candidate = parent / "config" / "routing-policy.yaml"
            if candidate.exists():
                return candidate
        raise FileNotFoundError("Unable to locate config/routing-policy.yaml for Velaris routing policy.")

    def _load_policy(self, policy_path: Path) -> dict[str, Any]:
        data = yaml.safe_load(policy_path.read_text(encoding="utf-8"))
        if not isinstance(data, dict):
            raise ValueError(f"Invalid routing policy file: {policy_path}")
        return data

    def _build_routing_context(self, plan: dict[str, Any], query: str) -> dict[str, Any]:
        scenario = str(plan.get("scenario", "general") or "general")
        governance = dict(plan.get("governance", {}))
        constraints = dict(plan.get("constraints", {}))
        capabilities = set(plan.get("capabilities", []))

        write_code = self._constraint_bool(constraints, "write_code", ("capability_demand", "writeCode"), default=False)
        # 默认外部副作用：高风险场景（如 robotclaw/procurement）需要审计
        _default_external_side_effects = self._registry.get_risk_level(scenario) == "high"
        external_side_effects = self._constraint_bool(
            constraints,
            "external_side_effects",
            ("capability_demand", "externalSideEffects"),
            default=_default_external_side_effects,
        )
        requires_audit = self._constraint_bool(
            constraints,
            "requires_audit",
            ("governance", "requiresAuditTrail"),
            default=bool(governance.get("requires_audit", False)),
        )
        task_complexity = self._constraint_str(
            constraints,
            "task_complexity",
            ("state", "taskComplexity"),
            default=self._infer_task_complexity(
                scenario=scenario,
                write_code=write_code,
                external_side_effects=external_side_effects,
                capability_count=len(capabilities),
            ),
        )
        risk_level = self._constraint_str(
            constraints,
            "risk_level",
            ("risk", "level"),
            default=self._infer_risk_level(
                scenario=scenario,
                requires_audit=requires_audit,
                external_side_effects=external_side_effects,
            ),
        )

        return {
            "goal": {
                "scenario": scenario,
                "query": query,
            },
            "risk": {
                "level": risk_level,
            },
            "state": {
                "taskComplexity": task_complexity,
            },
            "capabilityDemand": {
                "writeCode": write_code,
                "externalSideEffects": external_side_effects,
                "write": "contract_form" in capabilities,
            },
            "governance": {
                "requiresAuditTrail": requires_audit,
            },
        }

    def _constraint_bool(self, constraints: dict[str, Any], flat_key: str, nested_key: tuple[str, str], default: bool) -> bool:
        if flat_key in constraints:
            return bool(constraints[flat_key])
        parent = constraints.get(nested_key[0])
        if isinstance(parent, dict) and nested_key[1] in parent:
            return bool(parent[nested_key[1]])
        return default

    def _constraint_str(self, constraints: dict[str, Any], flat_key: str, nested_key: tuple[str, str], default: str) -> str:
        if flat_key in constraints and constraints[flat_key]:
            return str(constraints[flat_key])
        parent = constraints.get(nested_key[0])
        if isinstance(parent, dict) and nested_key[1] in parent and parent[nested_key[1]]:
            return str(parent[nested_key[1]])
        return default

    def _infer_task_complexity(self, scenario: str, write_code: bool, external_side_effects: bool, capability_count: int) -> str:
        if write_code or external_side_effects:
            return "complex"
        # 使用 registry 查询 risk_level 替代硬编码场景判断
        spec = self._registry.get(scenario)
        risk_level = spec.risk_level if spec else "medium"
        if capability_count >= 4 and risk_level == "high":
            return "complex"
        if risk_level == "low":
            return "simple"
        return "medium"

    def _infer_risk_level(self, scenario: str, requires_audit: bool, external_side_effects: bool) -> str:
        if requires_audit:
            return "high"
        if external_side_effects:
            return "medium"
        # 使用 registry 查询 risk_level 替代硬编码场景名判断
        return self._registry.get_risk_level(scenario)

    def _matches(self, condition: dict[str, Any], context: dict[str, Any]) -> bool:
        if "all" in condition:
            return all(self._matches(item, context) for item in condition["all"])
        if "any" in condition:
            return any(self._matches(item, context) for item in condition["any"])
        actual = self._read_path(context, str(condition["field"]))
        return self._compare(actual, str(condition["op"]), condition.get("value"))

    def _read_path(self, data: dict[str, Any], path: str) -> Any:
        cursor: Any = data
        for segment in path.split("."):
            if not isinstance(cursor, dict) or segment not in cursor:
                return None
            cursor = cursor[segment]
        return cursor

    def _compare(self, actual: Any, operator: str, expected: Any) -> bool:
        if operator == "eq":
            return actual == expected
        if operator == "ne":
            return actual != expected
        if operator == "gt":
            return isinstance(actual, (int, float)) and isinstance(expected, (int, float)) and actual > expected
        if operator == "gte":
            return isinstance(actual, (int, float)) and isinstance(expected, (int, float)) and actual >= expected
        if operator == "lt":
            return isinstance(actual, (int, float)) and isinstance(expected, (int, float)) and actual < expected
        if operator == "lte":
            return isinstance(actual, (int, float)) and isinstance(expected, (int, float)) and actual <= expected
        if operator == "in":
            return isinstance(expected, list) and actual in expected
        if operator == "not_in":
            return isinstance(expected, list) and actual not in expected
        return False

    def _compute_confidence(self, selected_rule: dict[str, Any] | None, sorted_rules: list[dict[str, Any]]) -> float:
        if selected_rule is None or not sorted_rules:
            return 0.55
        selected_id = str(selected_rule["id"])
        for index, rule in enumerate(sorted_rules):
            if str(rule["id"]) == selected_id:
                normalized = 1 - index / max(len(sorted_rules), 1)
                return round(0.6 + normalized * 0.4, 3)
        return 0.6
